### PR TITLE
Open up delayed/conditional logging

### LINKED ISSUE #227: Delayed logging
Right now logbook makes two decisions by default:
1. Requests and responses are logged the moment they are happening
2. Either both, request and response, are logged nor none of them

A common use case would be to deviate from decision 1 and allow users to delay logging of request until response is available. This is show cased in https://github.com/zalando/logbook/blob/0fd12f2fe8a30147fee2e77cfc60c206929c31f3/logbook-core/src/test/java/org/zalando/logbook/DelayedResponseLogWriter.java. This should be supported more easily.

### PR DESCRIPTION
**Pull request by whiskeysierra**
Fixes #227

### CODE REVIEW

File: logbook-core/src/main/java/org/zalando/logbook/ConditionalLogWriter.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+import java.io.IOException;
+import java.util.function.Predicate;
+
+import static org.apiguardian.api.API.Status.EXPERIMENTAL;
+
+@API(status = EXPERIMENTAL)
+public final class ConditionalLogWriter implements HttpLogWriter {
+
+    private final Predicate<? super HttpResponse> predicate;
```
[the remark this answers is not in the document]
Comment (whiskeysierra) [reply logbook-core/src/main/java/org/zalando/logbook/ConditionalLogWriter.java]: That's what I had initially, but it felt rather clumsy. Any good idea how the API could look like. Especially the one in `Conditions`?

[the remark this answers is not in the document]
Comment (AlexanderYastrebov) [reply logbook-core/src/main/java/org/zalando/logbook/ConditionalLogWriter.java]: Then maybe just make it very specific, e.g. `ResponseStatusAwareLogWriter`, use it as an example and let client define its own decorator the same way if needed?..

[the remark this answers is not in the document]
Comment (whiskeysierra) [reply logbook-core/src/main/java/org/zalando/logbook/ConditionalLogWriter.java]: I decided to go with `Predicate<Correlation<String, String>>`.

**Comment by AlexanderYastrebov** [comment]
:+1:

**Comment by whiskeysierra** [comment]
:-1:

**Comment by whiskeysierra** [comment]
https://github.com/zalando/logbook/issues/227#issuecomment-392973259 raises a valid concern. I'm more inclined to not adding any of this, just have recipes or examples available for people to implement this themselves. All of those scenarios (don't log 2xx at all, log 2xx but without body, ...) can all be implemented with a couple lines of code.

**Comment by whiskeysierra** [comment]
Closed in favor of a clean solution as part of 2.x, see #296

**Pull request closed without merging**